tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,210 @@
+                raise ValueError("Given data type does not match header program type.")
+        except KeyError as e:
+            logger.error('Ran into issue with header type: %s', e)
+
+        validator = Validator(get_header_schema())
+        is_valid = validator.validate(header)
+        logger.debug(validator.errors)
+
+        return is_valid, validator
+
+    except KeyError as e:
+        logger.error('Exception splicing header line or referencing it, please see line and error.')
+        logger.error(line)
+        logger.error(e)
+        return False, e
+
+def validate_trailer(line):
+    """Validate the trailer (last) line in a datafile."""
+    logger.info('Validating trailer line.')
+    validator = Validator(get_trailer_schema())
+
+    trailer = {
+        'title':        line[0:7],
+        'record_count': line[7:14],
+        'blank':        line[14:23],
+    }
+
+    is_valid = validator.validate(trailer)
+
+    logger.debug("Trailer title: '%s'", trailer['title'])
+    logger.debug("Trailer record count: '%s'", trailer['record_count'])
+    logger.debug("Trailer blank: '%s'", trailer['blank'])
+    logger.debug("Trailer errors: '%s'", validator.errors)
+
+    return is_valid, validator
+
+def get_header_line(datafile):
+    """Alters header line into string."""
+    # intentionally only reading first line of file
+    datafile.seek(0)
+    line = datafile.readline()
+    datafile.seek(0)  # reset file pointer to beginning of file
+
+    # datafile when passed via redis/db is a FileField which returns bytes
+    if isinstance(line, bytes):
+        line = line.decode()
+
+    line = line.strip()
+
+    if get_record_type(line) != 'HE':
+        return False, {'preparsing': 'First line in file is not recognized as a valid header.'}
+    elif len(line) != 23:
+        logger.debug("line: '%s' len: %d", line, len(line))
+        return False, {'preparsing': 'Header length incorrect.'}
+
+    return True, line
+
+def get_trailer_line(datafile):
+    """Alters the trailer line into usable string."""
+    # certify/transform input line to be correct form/type
+
+    # Don't want to read whole file, just last line, only possible with binary
+    # Credit: https://openwritings.net/pg/python/python-read-last-line-file
+    # https://stackoverflow.com/questions/46258499/
+    try:
+        datafile.seek(-2, os.SEEK_END)  # Jump to the second last byte.
+        while datafile.read(1) != b'\n':  # Check if new line.
+            datafile.seek(-2, os.SEEK_CUR)   # Jump two bytes back
+    except OSError:  # Either file is empty or contains one line.
+        datafile.seek(0)
+        return False, {'preparsing': 'File too short or missing trailer.'}
+
+    # Having set the file pointer to the last line, read it in.
+    line = datafile.readline().decode()
+    datafile.seek(0)  # Reset file pointer to beginning of file.
+
+    logger.info("Trailer line: '%s'", line)
+    if get_record_type(line) != 'TR':
```

Comment:
@andrew-jameson if we want to keep this validation check, i think we just need to ensure that there is only one record (at the end of the file) that has a length >= 14. some of the grantees dont add the 9 trailing blanks to the trailer, and its unclear why TDRS fusses only sometimes.

Reply:
Interesting, I hadn't thought about multiple header/trailer lines. At the moment, the parser would accept a file like so:

```
HEADER####...
HEADER####...
TRAILER####...
TRAILER####...
```

Reply:
yes, we'd want to reject something like that. to my knowledge, a couple of states include multiple headers and trailers in their file but it looks something like the below. it's because they are including multiple quarters of data in the file. we'd like them to stop doing this because they are overwriting data that was considered complete. 

```
HEADER
T1...
T2....
T3.....
....
....
TRAILER
HEADER
T1....
T2....
T3....
TRAILER
HEADER
....
TRAILER
```

Reply:
Can you add sample bad files like this to that same Sharepoint folder? Would love to increase our data set for tests.

Reply:
sure! its currently in the original bad file, but im currently creating a much shorter version for qasp review.